# Week-11: Tutorial on Pandas (Continued)

<font size='4'>

This week, we will continue learning `pandas`.
Before delving into new lecture notes, I will revisit Quiz 3.
Comments related to your final project task 1 have been posted, please address them accordingly.

In [54]:
# 0.1
import os
import glob
import numpy as np
import pandas as pd
print(os.getcwd())

/Users/tma33/Library/CloudStorage/OneDrive-EmoryUniversity/Emory/Rollins SPH/2025/BIOS-584/python_proj


In [55]:
ptsd_dir = '{}/data/PTSD dataset.xlsx'.format(os.getcwd())
ptsd_df = pd.read_excel(ptsd_dir, sheet_name='main_dataset')
# print(ptsd_df.columns)

## 5. Cleaning data using pandas 

<font size='4'>

* Data cleaning is one of the most common but important tasks in data science.
* Pandas allows you to preprocess data for multiple uses, including but not limited to training machine learning and deep learning models.
* Always check the missingness of the dataset first!

In [56]:
# 5.0.1
print(ptsd_df.isnull().sum().sum())

48920


### 5.1. Dropping missing values
<font size='4'>
    
* One way to deal with missing data is to simply drop it.
* This may be useful when you have plenty of data and losing a small portion won't impact the downstream analysis.
* You can use a `.dropna()` method.
* As an example, we apply this method to a copy of original dataset.
* https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html

In [57]:
# 5.1.1
ptsd_df_2 = ptsd_df.copy()
print(ptsd_df_2.shape) # before dropping 
ptsd_df_2 = ptsd_df_2.dropna()
print(ptsd_df_2.shape) # after dropping

(483, 439)
(0, 439)


<font size='4'>

* The resulting `ptsd_df_2` ends up with no rows because `dropna()` will remove the entire row as long as there exists one missing value.
* Let's look at the dataset with its first eight columns.

In [58]:
# 5.1.2
ptsd_df_3 = ptsd_df.copy()
ptsd_df_3 = ptsd_df_3.iloc[:, range(8)]
print(ptsd_df_3.head(n=3))
print(ptsd_df_3.isnull().sum())
print(ptsd_df_3.shape) # before dropping
ptsd_df_3 = ptsd_df_3.dropna()
print(ptsd_df_3.shape) # after dropping

   record_id  ptsdpresent_caps  caps_minuspcl  caps_minuspcl_code  \
0        460                 1            -56                   1   
1       5942                 1            -49                   1   
2       5366                 1            -48                   1   

   caps_intake  pcl5_score_intake  pcl5month_score.baseline  \
0           15                 71                      75.0   
1           23                 72                       NaN   
2           31                 79                      62.0   

   pcl5week_score.completion  
0                       40.0  
1                       52.0  
2                       23.0  
record_id                     0
ptsdpresent_caps              0
caps_minuspcl                 0
caps_minuspcl_code            0
caps_intake                   0
pcl5_score_intake             0
pcl5month_score.baseline      9
pcl5week_score.completion    27
dtype: int64
(483, 8)
(450, 8)


<font size='4'>

* The original sample size is reduced from 483 to 450.
* When one variable has the missingness smaller than **10%**, it is okay to simply remove them.

In [59]:
# 5.1.3
print(ptsd_df.shape)
ptsd_df_4 = ptsd_df.copy()
ptsd_df_4.dropna(inplace=True, axis=1)
print(ptsd_df_4.shape)
ptsd_df_4.head(n=3)

(483, 439)
(483, 71)


record_id  ptsdpresent_caps  caps_minuspcl  caps_minuspcl_code  \
0        460                 1            -56                   1   
1       5942                 1            -49                   1   
2       5366                 1            -48                   1   

   caps_intake  pcl5_score_intake  mdd_code  sexual_trauma  sud_code  age_iop  \
0           15                 71         0              0         0       39   
1           23                 72         0              1         0       36   
2           31                 79         1              0         1       42   

   ...  primary_diagnosis_icd  secondary_icd_dx_code_1  \
0  ...                 F43.12                    F32.1   
1  ...                 F43.10                    F41.0   
2  ...                 F43.10                   F10.20   

                  fullcaps  gender_src  marital_status  number_of_children  \
0   missing all items 1-20      Female        Divorced                   0   
1  all items 1-20 complete      Female         Married                   0   
2  all items 1-20 complete        Male         Married                   1   

     branch_src  military_service_status_src  state_code  \
0       US Army                  Active Duty          TX   
1  US Air Force                      Retired          GA   
2       US Army                  Active Duty          GA   

   va_benefits_indicator  
0                    Yes  
1                    Yes  
2                      N  

[3 rows x 71 columns]

<font size='4'>
    
* The `axis` parameter lets you specify whether you are dropping rows, or columns, with missing values.
    * The default `axis` removes the rows containing `NaN`.
    * For a two-dimensional array, use `axis=1` to remove the columns with one or more `NaN` values.
    * `inplace=True` lets you skip saving the output of `.dropna()` into a new DataFrame.
* In this case, the column number is reduced from 439 to 71.

* Of couse, we can drop both rows and columns with missing values by setting the `how` parameter.
    * `any`: If any missing values are present, drop that row or column.
    * `all`: If all values are NA, drop that row or column.

In [60]:
# 5.1.4
ptsd_df_5 = ptsd_df.copy()
ptsd_df_5.dropna(inplace=True, how='all')
print(ptsd_df_5.shape)
ptsd_df_5.head()
# In this case, when we set `all`, 
# nothing is reduced because there is no completely missing row.

(483, 439)


record_id  ptsdpresent_caps  caps_minuspcl  caps_minuspcl_code  \
0        460                 1            -56                   1   
1       5942                 1            -49                   1   
2       5366                 1            -48                   1   
3       3480                 1            -47                   1   
4       1029                 1            -46                   1   

   caps_intake  pcl5_score_intake  pcl5month_score.baseline  \
0           15                 71                      75.0   
1           23                 72                       NaN   
2           31                 79                      62.0   
3           31                 78                      76.0   
4           28                 74                      52.0   

   pcl5week_score.completion  pcl5month_score.3_month_follow_up  mdd_code  \
0                       40.0                                NaN         0   
1                       52.0                               29.0         0   
2                       23.0                                NaN         1   
3                       72.0                                NaN         0   
4                       29.0                               36.0         0   

   ...  state_code  va_benefits_indicator  wwp_alumni_indicator  army  \
0  ...          TX                    Yes                    No   1.0   
1  ...          GA                    Yes                    No   0.0   
2  ...          GA                      N                     N   1.0   
3  ...          GA                    Yes                   Yes   1.0   
4  ...          GA                    Yes                    No   0.0   

   airforce  marines  navy  coastguard  nationalguard  reserve  
0       0.0      0.0   0.0         0.0            0.0      0.0  
1       1.0      0.0   0.0         0.0            0.0      0.0  
2       0.0      0.0   0.0         0.0            0.0      0.0  
3       0.0      0.0   0.0         0.0            0.0      0.0  
4       0.0      1.0   0.0         0.0            0.0      0.0  

[5 rows x 439 columns]

### 5.2. Replacing missing values

<font size='4'>

* When the missing percentage is moderate (>= 15% in my view), dropping values may lose information and introduce bias to your effect estimation.
* Replacing missing values with other values is preferred.
* You can fill in the missing values with a summary statistics, i.e., mean value, or apply some statistical methodology to infer a number, i.e., multiple imputation.
* Relevant pandas method is named `.fillna()`.
* https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html

In [61]:
# 5.2.1
ptsd_df_3 = ptsd_df.copy()
ptsd_df_3 = ptsd_df_3.iloc[:, range(8)]
print(ptsd_df_3.isnull().sum())
print()

pcl5_baseline_mean = ptsd_df_3['pcl5month_score.baseline'].mean()
print(pcl5_baseline_mean)
print()

# ptsd_df_3['pcl5month_score.baseline'].fillna(pcl5_baseline_mean, inplace=True)
# this syntax will be deprecated for Pandas 3.0
ptsd_df_3.fillna({"pcl5month_score.baseline": pcl5_baseline_mean}, inplace=True)
print(ptsd_df_3.isnull().sum())

record_id                     0
ptsdpresent_caps              0
caps_minuspcl                 0
caps_minuspcl_code            0
caps_intake                   0
pcl5_score_intake             0
pcl5month_score.baseline      9
pcl5week_score.completion    27
dtype: int64

51.164556962025316

record_id                     0
ptsdpresent_caps              0
caps_minuspcl                 0
caps_minuspcl_code            0
caps_intake                   0
pcl5_score_intake             0
pcl5month_score.baseline      0
pcl5week_score.completion    27
dtype: int64


In [62]:
# 5.2.2
# Assign the mean of pcl5week_score.completion to its 27 missing values.
# Write down your code.


### 5.3. Handling Duplicated Values

<font size='4'>

* Let's manually create some duplicates to the existing dataset.
* You can remove all duplicated rows (by default) from the DataFrame using `.drop_duplicates()` method.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html

In [63]:
# 5.3.1
print(ptsd_df_3.shape)
ptsd_df_3_dup = pd.concat([ptsd_df_3, ptsd_df_3])
print(ptsd_df_3_dup.shape)

(483, 8)
(966, 8)


In [64]:
# 5.3.2
ptsd_df_3_no_dup = ptsd_df_3_dup.drop_duplicates()
ptsd_df_3_no_dup.shape

(483, 8)

### 5.4. Renaming Columns

<font size='4'>

* You can use `.rename()` method to modify the column names.
* For example, we want to change `pcl5month_score.baseline` to `pcl5_score_baseline` and `pcl5week_score.completion` to `pcl5_score_completion` in `ptsd_df_3`.

In [65]:
# 5.4.1
print(ptsd_df_3.columns)
ptsd_df_3.rename(columns={'pcl5month_score.baseline':'pcl5_score_baseline', 
                          'pcl5week_score.completion':'pcl5_socre_completion'}, inplace=True)
print(ptsd_df_3.columns)

Index(['record_id', 'ptsdpresent_caps', 'caps_minuspcl', 'caps_minuspcl_code',
       'caps_intake', 'pcl5_score_intake', 'pcl5month_score.baseline',
       'pcl5week_score.completion'],
      dtype='object')
Index(['record_id', 'ptsdpresent_caps', 'caps_minuspcl', 'caps_minuspcl_code',
       'caps_intake', 'pcl5_score_intake', 'pcl5_score_baseline',
       'pcl5_socre_completion'],
      dtype='object')


<font size='4'>

* You can also directly assign column names as a list to the DataFrame.
* Make sure the variable order in the list is consistent with the column names.

In [66]:
# 5.4.2
ptsd_df_3.columns = ['record_id', 'ptsd_present_caps', 'caps_minus_pcl_num', 'caps_minus_pcl_code',
                    'caps_score_intake', 'pcl5_score_intake', 'pcl5_score_baseline', 'pcl5_score_completion']
ptsd_df_3.head(n=3)

record_id  ptsd_present_caps  caps_minus_pcl_num  caps_minus_pcl_code  \
0        460                  1                 -56                    1   
1       5942                  1                 -49                    1   
2       5366                  1                 -48                    1   

   caps_score_intake  pcl5_score_intake  pcl5_score_baseline  \
0                 15                 71            75.000000   
1                 23                 72            51.164557   
2                 31                 79            62.000000   

   pcl5_score_completion  
0                   40.0  
1                   52.0  
2                   23.0

<font size='4'>

* For more details, please read this checklist: https://www.datacamp.com/blog/infographic-data-cleaning-checklist

## 6. Data Analysis in Pandas

### 6.1. Summary Statistics (mean, median, and mode)

<font size='4'>
    
* `.mean()` for mean
* `.median()` for median
* `.mode()` for mode

* Similar to `np.mean(), np.median()` functions, Pandas has three methods to compute mean, median, and mode for the DataFrame.

In [67]:
# 6.1.1
ptsd_df_3.mean()
# However, this only applies to continuous or ordinal columns. You need to check your results carefully.

record_id                3184.631470
ptsd_present_caps           1.000000
caps_minus_pcl_num        -17.532091
caps_minus_pcl_code         1.983437
caps_score_intake          36.080745
pcl5_score_intake          53.612836
pcl5_score_baseline        51.164557
pcl5_score_completion      29.438596
dtype: float64

### 6.2. Create new columns based on existing columns

<font size='4'>

* Similar to R, pandas can easily create a new column using data from existing columns.
* For example, let's create a new column `pcl_5_mean` by taking the average of `pcl_5_score_intake`, `pcl_5_score_baseline`, and `pcl_5_score_completion`.
    * This value is not clinically meaningful.

In [68]:
# 6.2.1
ptsd_df_3['pcl5_mean'] = (ptsd_df_3['pcl5_score_intake'] + ptsd_df_3['pcl5_score_baseline'] + ptsd_df_3['pcl5_score_completion'])/3
print(ptsd_df_3['pcl5_mean'])
ptsd_df_3.head(n=3)

0      62.000000
1      58.388186
2      54.666667
3      75.333333
4      51.666667
         ...    
478    30.000000
479    26.333333
480          NaN
481    24.666667
482    50.666667
Name: pcl5_mean, Length: 483, dtype: float64


record_id  ptsd_present_caps  caps_minus_pcl_num  caps_minus_pcl_code  \
0        460                  1                 -56                    1   
1       5942                  1                 -49                    1   
2       5366                  1                 -48                    1   

   caps_score_intake  pcl5_score_intake  pcl5_score_baseline  \
0                 15                 71            75.000000   
1                 23                 72            51.164557   
2                 31                 79            62.000000   

   pcl5_score_completion  pcl5_mean  
0                   40.0  62.000000  
1                   52.0  58.388186  
2                   23.0  54.666667

### 6.3. Counting using `.value_counts()`

<font size='4'>

* For categorical variables, we use `.value_counts()` method
* https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.value_counts.html
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html

In [69]:
# 6.3.1
print(ptsd_df_3['caps_minus_pcl_code'].value_counts())
print(ptsd_df_3['caps_minus_pcl_code'].value_counts(normalize=True))
pd.concat([ptsd_df_3['caps_minus_pcl_code'].value_counts(), 
           ptsd_df_3['caps_minus_pcl_code'].value_counts(normalize=True)*100], axis=1)
# with a better format

caps_minus_pcl_code
2    325
1     83
3     75
Name: count, dtype: int64
caps_minus_pcl_code
2    0.672878
1    0.171843
3    0.155280
Name: proportion, dtype: float64


count  proportion
caps_minus_pcl_code                   
2                      325   67.287785
1                       83   17.184265
3                       75   15.527950

### 6.4. Aggregating data with `.groupby()` in pandas

<font size='4'>

* Pandas allows you to aggregate values by grouping them by specific column values using `.groupby()` method.
* Pay attention to the order of your methods.
* Use `[]` to include multiple variables.

In [70]:
# 6.4.1
ptsd_df_3.groupby('caps_minus_pcl_code')['pcl5_score_baseline'].mean()

caps_minus_pcl_code
1    60.100351
2    51.519455
3    39.737722
Name: pcl5_score_baseline, dtype: float64

In [71]:
# 6.4.2
pcl5_name_ls = ['pcl5_score_baseline', 'pcl5_score_completion']
ptsd_df_3.groupby('caps_minus_pcl_code')[pcl5_name_ls].mean()

pcl5_score_baseline  pcl5_score_completion
caps_minus_pcl_code                                            
1                              60.100351              35.425000
2                              51.519455              29.771242
3                              39.737722              21.142857

### 6.5. Pivot tables

<font size='4'>

* Pandas enables you to calculate summary statistics as pivot tables.
* Use `pandas.pivot_table()` function
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html
* `index` is the row variable, `columns` is  the column variable, `values` are the outcome of interest after aggregation.

In [72]:
# 6.5.1
pd.pivot_table(ptsd_df, values='pcl5week_score.completion', index='caps_minuspcl_code', columns=['gender_code'])

gender_code                 1          2
caps_minuspcl_code                      
1                   34.765957  36.363636
2                   30.659341  28.467742
3                   21.422222  20.640000

In [73]:
# 6.5.2
# You can examine multiple outcomes.
pd.pivot_table(ptsd_df, values=['pcl5week_score.completion', 'pcl5month_score.baseline'], index='caps_minuspcl_code', columns='gender_code')

pcl5month_score.baseline             \
gender_code                               1          2   
caps_minuspcl_code                                       
1                                 58.458333  63.030303   
2                                 51.541667  51.500000   
3                                 40.500000  37.360000   

                   pcl5week_score.completion             
gender_code                                1          2  
caps_minuspcl_code                                       
1                                  34.765957  36.363636  
2                                  30.659341  28.467742  
3                                  21.422222  20.640000

In [74]:
# 6.5.3
pd.pivot_table(ptsd_df, values=['pcl5week_score.completion', 'pcl5month_score.baseline'], index='gender_code', columns='caps_minuspcl_code')

pcl5month_score.baseline                    \
caps_minuspcl_code                        1          2      3   
gender_code                                                     
1                                 58.458333  51.541667  40.50   
2                                 63.030303  51.500000  37.36   

                   pcl5week_score.completion                        
caps_minuspcl_code                         1          2          3  
gender_code                                                         
1                                  34.765957  30.659341  21.422222  
2                                  36.363636  28.467742  20.640000